file-upload.md

```diff
@@ -0,0 +1,304 @@
+# Uploading files from an HTML form
+
+When uploading a file from an HTML form, the file content is not sent as the request body.
+This is because an HTML form may contain more than one input, and in a POST request all input values will be sent in the request body.
+A form which has an `<input type="file">` must use `enctype="multipart/form-data"`, according to the [specification](https://www.ietf.org/rfc/rfc1867.txt).
+The file content can be sent as raw data, or base64 encoded, according to the [OpenAPI specification](https://swagger.io/docs/specification/describing-request-body/file-upload/):
+"Files use a `type: string` schema with `format: binary` or `format: base64`, depending on how the file contents will be encoded."
+
+How this all works when using Roaster is illustrated in the examples below.
+
+## Text file upload
+
+In this example, the filename is sent as a separate parameter.
+This is not strictly necessary, but it shows how to send a filename which may be different from the original filename as part of the URL path.
+
+First look at the HTML form.
+
+```html
+    <form action="#" method="POST" enctype="multipart/form-data" id="textFileUploadForm"
+      onsubmit="return submitTextFileUpload()"
+    >
+      <fieldset>
+        <legend>Upload a text file.</legend>
+        <p><input type="file" name="file" id="textFileUploadInput"/></p>
+        <p><input type="submit" value="Upload text file"/></p>
+      </fieldset>
+    </form>
+```
+
+The javascript function used for upload:
+
+```javascript
+      /**
+       * Upload a text file.
+       * In this example, the filename is part of the URL, so the form action is modified.
+       * The filename is also sent in the form-data, so this is not really necessary.
+       */
+      function submitTextFileUpload()
+      {
+        const fileName = document.getElementById('textFileUploadInput').files[0].name;
+        document.getElementById('textFileUploadForm').action = `../upload-text/${fileName}`;
+        // Return true to submit immediately.
+        return true;
+      }
+```
+The form is at the URL `static/upload.html`, which is why the relative URL for uploading is `../upload-text/${fileName}`.
+
+The OpenAPI specification for the upload is:
+
+```json:
```

Comment:
Looks good! Except there shouldn't be a colon (`:`) after the language code here.